apache/druid
# Add Integration Test for functionality of kinesis ingestion

Add Integration Test for functionality of kinesis ingestion

### Description
The new set of integration test for Kinesis follows the same concept as Bring Your Own Cloud (BYOC) that S3, GCS, Azure Integration tests uses (https://github.com/apache/druid/pull/9501). Basically, anyone running will have to provide their own Kinesis credentials in a conf file and pass the file to mvn using -Doverride.config.path

Added following Integration Test for functionality of kinesis ingestion:

1. Functional tests when Druid and Kafka are in stable state
- legacy parser
- inputFormat
- Greater than 1 taskCount

2. Functional tests when Druid is in an unstable state 
- losing nodes
- Stop/start supervisor

3. Functional tests when Kafka is in an unstable state 
- adding partitions
- removing partitions

To verify ingestion:
- Kafka lag should be minimal, the consumer should be able to pull off the queue at a comparable rate to the producer.
- Realtime queries works from the indexing tasks
- Queries works reading from historical segments (after handed off)
- Queries return expected count/value/etc.

Added integration test infrastructure/helper:
- Data generator - Generate data to kinesis stream (Kafka can also be refactor to use this later)
- DruidAdminClient - to control the Integration test's Druid Docker cluster
- KinesisAdminClient - Create stream, delete stream, etc.
- Since tests can fail or be killed un-expectedly, added druid-ci-expire-after tag to all stream created by integration test. The value to this tag is a timestamp that can be used by a lambda function to remove unused stream. (help make clean up easier)
- Also made debugging integration test easier by automatically enabling debug mode and exposing debug port on integration test's druids docker processes

This PR has:
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
